### ISSUE TITLE
AddressSanitizer: global-buffer-overflow at nmd_assembly.h:9257 in nmd_x86_format

### ISSUE THREAD

**Issue by fuzzah** [Nomade040/nmd#9]
Hello!
 
With help of fuzzing I've found out that there exist unchecked **op** value of 0x16 in function **nmd_x86_format** (file nmd_assembly.h, line 9257).
**_NMD_C(op)** results in 6 which is not valid index for array **prefixF2_mnemonics** of size 3 (thus causing buffer overflow on further read of **source** pointer in function **_nmd_append_string**).
 
Two possible (naive?) fixes:
1) check for **op** to not equal 0x16 before calling to **_nmd_append_string** (would break some logic?);
2) initialize arrays like **prefixF2_mnemonics** with some zeroes (NULL pointers) to have size of at least 8 elements each (because ``op >= 0x10 && op <= 0x17``) AND in function **_nmd_append_string** check **source** argument for NULL before while-loop.

Full ASAN message:
 
```
==2055366==ERROR: AddressSanitizer: global-buffer-overflow on address 0x5555557d4ef0 at pc 0x555555737e4d bp 0x7fffffffd950 sp 0x7fffffffd948
READ of size 8 at 0x5555557d4ef0 thread T0
    #0 0x555555737e4c in nmd_x86_format /home/user/git/nmd/./nmd_assembly.h:9257:29
    #1 0x55555578c74e in main /home/user/git/nmd/nmd_fuzz.c:64:13
    #2 0x7ffff7c76b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)
    #3 0x55555558e28d in _start (/home/user/git/nmd/nmd_fuzz_asan+0x3a28d)

0x5555557d4ef0 is located 16 bytes to the left of global variable '__const.nmd_x86_format.no_prefix_mnemonics' defined in 'nmd_fuzz.c' (0x5555557d4f00) of size 64
0x5555557d4ef0 is located 24 bytes to the right of global variable '__const.nmd_x86_format.prefixF2_mnemonics' defined in 'nmd_fuzz.c' (0x5555557d4ec0) of size 24
SUMMARY: AddressSanitizer: global-buffer-overflow /home/user/git/nmd/./nmd_assembly.h:9257:29 in nmd_x86_format
Shadow bytes around the buggy address:
  0x0aab2aaf2980: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0aab2aaf2990: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0aab2aaf29a0: 00 00 00 00 00 00 f9 f9 f9 f9 f9 f9 f9 f9 f9 f9
  0x0aab2aaf29b0: 00 00 00 00 00 00 00 00 f9 f9 f9 f9 00 00 00 f9
  0x0aab2aaf29c0: f9 f9 f9 f9 00 00 00 00 f9 f9 f9 f9 00 00 00 00
=>0x0aab2aaf29d0: 00 00 00 00 f9 f9 f9 f9 00 00 00 f9 f9 f9[f9]f9
  0x0aab2aaf29e0: 00 00 00 00 00 00 00 00 f9 f9 f9 f9 00 00 00 00
  0x0aab2aaf29f0: 00 00 00 00 00 00 00 00 00 00 00 00 f9 f9 f9 f9
  0x0aab2aaf2a00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0aab2aaf2a10: f9 f9 f9 f9 00 00 00 00 00 00 00 00 f9 f9 f9 f9
  0x0aab2aaf2a20: 00 00 00 00 00 00 00 00 f9 f9 f9 f9 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==2055366==ABORTING
```

To reproduce this behavior one may change buffer contents in examples/assembly_example.c to contents of this file and compile either without optimization (e.g. -O0), or with address sanitizer (-fsanitize=address). Note that I used clang to compile it.

[global-buffer-overflow_nmd_assembly_h__line_9257.tar.gz](https://github.com/Nomade040/nmd/files/6137502/global-buffer-overflow_nmd_assembly_h__line_9257.tar.gz)

**Comment by Nomade040** [comment]
Thank you for using your time to find this issue.

I confirmed that the following sequence of bytes causes the formatter to access bad memory.
`F3 F2 0F 16 29 15 16 04 E8 10 20 22`

The sequence may be reduced to `F3 F2 0F 16 29` according to the format of an instruction in the x86 architecture.

In short, the sequence generated by the fuzzer would probably not be emitted by an assembler because the prefixes `F3` and `F2` should be mutually exclusive in this context. However, an instruction should not be considered invalid when the prefixes are organized in this way, because the last prefix is considered to be the effective one.

I think the formatter should not perform instruction validation, because it is the job of the decoder. The decoder(`nmd_x86_decode()`) does actually have checks for this case:

```
else if (instruction->prefixes & (NMD_X86_PREFIXES_REPEAT | NMD_X86_PREFIXES_REPEAT_NOT_ZERO))
{
    if ((op >= 0x13 && op <= 0x17 && !(op == 0x16 && instruction->prefixes & NMD_X86_PREFIXES_REPEAT)) || ...)
        return false;
}
```
The code snippet checks for non-existing instructions in the 2-byte opcode map as the following image taken from the intel's x86 SDM shows:
![image](https://user-images.githubusercontent.com/31376198/111162427-e63a0b00-857a-11eb-811c-71ba28afec38.png)

To do that, it checks if the instruction has the prefix `F3`(NMD_X86_PREFIXES_REPEAT) or `F2`(NMD_X86_PREFIXES_REPEAT_NOT_ZERO), and is within the range 13-17. It also checks for an exception which is the instruction that has the prefix `F3` and opcode `16` because that instruction exists.

Basically the opcode of the invalid instruction in the sequence is `0x16` which passes the check `op >= 0x13 && op <= 0x17` and it should also have passed the second condition(given that its effective prefix is `F2`) so the decoder would return false and the formatter would not even have been called in the first place. The variable `instruction->prefixes` is a mask of all prefixes present in the instruction, however when checking for the exception we only actually want the effective prefix which is the prefix closer to the opcode(the rightmost one), so the code should have used the variable `instruction->simd_prefix` instead. The invalid instruction tricks the decoder into thinking it is the exception instruction it that condition.

**Comment by Nomade040** [comment]
Issue fixed by commit 6543c592c33b977be9797a264f85130bb7a53a56.
